solo/utils/whitening.py

```diff
@@ -66,3 +68,167 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
         decorrelated = conv2d(xn, inv_sqrt)
 
         return decorrelated.squeeze(2).squeeze(2)
+
+
+class iterative_normalization_py(torch.autograd.Function):
+    @staticmethod
+    def forward(ctx, *args) -> torch.Tensor:
+        X, running_mean, running_wmat, nc, ctx.T, eps, momentum, training = args
+
+        # change NxCxHxW to (G x D) x(NxHxW), i.e., g*d*m
+        ctx.g = X.size(1) // nc
+        x = X.transpose(0, 1).contiguous().view(ctx.g, nc, -1)
+        _, d, m = x.size()
+        saved = []
+        if training:
+            # calculate centered activation by subtracted mini-batch mean
+            mean = x.mean(-1, keepdim=True)
+            xc = x - mean
+            saved.append(xc)
+            # calculate covariance matrix
+            P = [None] * (ctx.T + 1)
+            P[0] = torch.eye(d).to(X).expand(ctx.g, d, d)
+            Sigma = torch.baddbmm(
+                beta=eps, input=P[0], alpha=1.0 / m, batch1=xc, batch2=xc.transpose(1, 2)
+            )
+            # reciprocal of trace of Sigma: shape [g, 1, 1]
+            rTr = (Sigma * P[0]).sum((1, 2), keepdim=True).reciprocal_()
+            saved.append(rTr)
+            Sigma_N = Sigma * rTr
+            saved.append(Sigma_N)
+            for k in range(ctx.T):
+                P[k + 1] = torch.baddbmm(
+                    beta=1.5,
+                    input=P[k],
+                    alpha=-0.5,
+                    batch1=torch.matrix_power(P[k], 3),
+                    batch2=Sigma_N,
+                )
+            saved.extend(P)
+            wm = P[ctx.T].mul_(
+                rTr.sqrt()
+            )  # whiten matrix: the matrix inverse of Sigma, i.e., Sigma^{-1/2}
+
+            running_mean.copy_(momentum * mean + (1.0 - momentum) * running_mean)
+            running_wmat.copy_(momentum * wm + (1.0 - momentum) * running_wmat)
```

Comment:
I'm getting this error. It's probably related to the `clone()` that I said we could remove. Sorry about that.

`E           RuntimeError: unsupported operation: more than one element of the written-to tensor refers to a single memory location. Please clone() the tensor before performing the operation.`

Reply:
I reintroduced the .clone().